drivers/regulator/regulator_axp2101.c

```diff
@@ -0,0 +1,573 @@
+/*
+ * Copyright (c) 2021 NXP
+ * Copyright (c) 2023 Martin Kiepfer <mrmarteng@teleschirm.org>
+ * Copyright (c) 2024 Deng Baoan <dengbaoan@foxmail.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT x_powers_axp2101_regulator
+
+#include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/regulator.h>
+#include <zephyr/sys/linear_range.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/logging/log_instance.h>
+
+LOG_MODULE_REGISTER(regulator_axp2101, CONFIG_REGULATOR_LOG_LEVEL);
+
+#define AXP2101_DC_ONOFF_DVM_CTRL (0x80)
+#define AXP2101_DC_WORKMODE_CTRL  (0x81)
+#define AXP2101_DC_VOL1_CTRL      (0x82)
+#define AXP2101_DC_VOL2_CTRL      (0x83)
+#define AXP2101_DC_VOL3_CTRL      (0x84)
+#define AXP2101_DC_VOL4_CTRL      (0x85)
+#define AXP2101_DC_VOL5_CTRL      (0x86)
+#define AXP2101_DC_OC_CTRL        (0x87)
+
+#define AXP2101_LDO_ONOFF_CTRL0 (0x90)
+#define AXP2101_LDO_ONOFF_CTRL1 (0x91)
+#define AXP2101_LDO_VOL_A1_CTRL (0x92)
+#define AXP2101_LDO_VOL_A2_CTRL (0x93)
+#define AXP2101_LDO_VOL_A3_CTRL (0x94)
+#define AXP2101_LDO_VOL_A4_CTRL (0x95)
+#define AXP2101_LDO_VOL_B1_CTRL (0x96)
+#define AXP2101_LDO_VOL_B2_CTRL (0x97)
+#define AXP2101_LDO_VOL_C1_CTRL (0x98)
+#define AXP2101_LDO_VOL_D1_CTRL (0x99)
+#define AXP2101_LDO_VOL_D2_CTRL (0x9A)
+
+/* DCDCs */
+#define AXP2101_DCDC_MODE_AUTO 0x00U
+#define AXP2101_DCDC_MODE_PWM  0x01U
+
+struct regulator_axp2101_desc {
+	const uint8_t enable_reg;
+	const uint8_t enable_mask;
+	const uint8_t enable_val;
+	const uint8_t vsel_reg;
+	const uint8_t vsel_mask;
+	const uint8_t vsel_bitpos;
+	const int32_t max_ua;
+	const uint8_t workmode_reg;
+	const uint8_t workmode_mask;
+	const uint8_t workmode_pwm_val;
+	const uint8_t num_ranges;
+	const struct linear_range *ranges;
+};
+
+struct regulator_axp2101_data {
+	struct regulator_common_data data;
+};
+
+struct regulator_axp2101_config {
+	struct regulator_common_config common;
+	const struct regulator_axp2101_desc *desc;
+	const struct device *mfd;
+	const struct i2c_dt_spec i2c;
+
+	LOG_INSTANCE_PTR_DECLARE(log);
+};
+
+/*! ----------------------------------------------------------------------------------------*/
```

Comment:
Comment style should be follow https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style